src/gui/callstatechecker.cpp

```diff
@@ -0,0 +1,181 @@
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QJsonArray>
+
+#include "callstatechecker.h"
+#include "account.h"
+
+namespace OCC {
+
+Q_LOGGING_CATEGORY(lcCallStateChecker, "nextcloud.gui.callstatechecker", QtInfoMsg)
+
+const int successStatusCode = 200;
+
+CallStateChecker::CallStateChecker(QObject *parent)
+    : QObject(parent)
+{
+    setup();
+}
+
+void CallStateChecker::setup()
+{
+    _notificationTimer.setSingleShot(true);
+    _notificationTimer.setInterval(60 * 1000);
+    connect(&_notificationTimer, &QTimer::timeout, this, &CallStateChecker::slotNotificationTimerElapsed);
+
+    _statusCheckTimer.setInterval(1000);
+    connect(&_statusCheckTimer, &QTimer::timeout, this, &CallStateChecker::slotStatusCheckTimerElapsed);
+
+    connect(this, &CallStateChecker::tokenChanged, this, &CallStateChecker::reset);
+    connect(this, &CallStateChecker::accountStateChanged, this, &CallStateChecker::reset);
+}
+
+QString CallStateChecker::token() const
+{
+    return _token;
+}
+
+void CallStateChecker::setToken(const QString &token)
+{
+    _token = token;
+    Q_EMIT tokenChanged();
+}
+
+AccountState* CallStateChecker::accountState() const
+{
+    return _accountState;
+}
+
+void CallStateChecker::setAccountState(AccountState *state)
+{
+    _accountState = state;
+    Q_EMIT accountStateChanged();
+}
+
+void CallStateChecker::startChecking()
+{
+    qCInfo(lcCallStateChecker, "Starting to check state of call.");
+    _notificationTimer.start();
+    _statusCheckTimer.start();
+}
+
+void CallStateChecker::stopChecking()
+{
+    qCInfo(lcCallStateChecker, "Stopping checking of call state.");
+    _notificationTimer.stop();
+    _statusCheckTimer.stop();
+    _stateCheckJob.clear();
+}
+
+void CallStateChecker::reset()
+{
+    qCInfo(lcCallStateChecker, "Resetting call check");
+    stopChecking();
+    startChecking();
+}
+
+void CallStateChecker::slotNotificationTimerElapsed()
+{
+    qCInfo(lcCallStateChecker, "Notification timer elapsed, stopping call checking");
+    stopChecking();
+    Q_EMIT notificationTimerElapsed();
+}
+
+void CallStateChecker::slotStatusCheckTimerElapsed()
+{
+    // Don't run check if another check is still ongoing
+    if (!_stateCheckJob.isNull()) {
+        return;
+    }
+
+    startCallStateCheck();
+}
+
+bool CallStateChecker::isAccountServerVersion22OrLater() const
+{
+    if(!_accountState || !_accountState->account()) {
+        return false;
+    }
+
+    const auto accountNcVersion = _accountState->account()->serverVersionInt();
+    const auto ncVersion22 = OCC::Account::makeServerVersion(22, 0, 0);
```

Comment:
could you try to make this a constexpr ?
looks like a good candidate

Reply:
`non constexpr function 'makeServerVersion' cannot be used in a constant expression` :(

Reply:
but cannot you make `OCC::Account::makeServerVersion` a `constexpr` ?

Reply:
In trying to do this I get this error:

```
auto components = serverVersion().split('.'); 

[account.cpp:654:10: Variable of non-literal type 'QStringList' cannot be defined in a constexpr function
qstringlist.h:111:7: 'QStringList' is not literal because it is not an aggregate and has no constexpr constructors other than copy or move constructors]
```

I there some other way we can get the major, min, micro numbers?

Reply:
let's maybe close this but I was talking about using this definition of `OCC::Account::makeServerVersion`
```
    static constexpr int makeServerVersion(int majorVersion, int minorVersion, int patchVersion)
    {
        return (majorVersion << 16) + (minorVersion << 8) + patchVersion;
    }
```
should then allow to get those constant versions be real constant

Reply:
but that is not really important to fix now
would be nice to have

Reply:
I was looking at the wrong method 🤦‍♂️ I was able to make it a constexpr now